sw/host/opentitanlib/src/test_utils/elf_debugger.rs

```diff
@@ -0,0 +1,337 @@
+        }
+        Ok(())
+    }
+
+    /// Assert the current program counter is within the given range.
+    pub fn expect_pc_range(&self, range: Range<impl Into<SymbolicAddress>>) -> Result<()> {
+        let start = self.resolve(range.start)?;
+        let end = self.resolve(range.end)?;
+        let pc = self.get_pc()?;
+        log::info!("PC = {:#x}, expected PC = {}..{}", pc, start, end,);
+        if !(start.resolution..end.resolution).contains(&pc) {
+            anyhow::bail!("unexpected PC");
+        }
+        Ok(())
+    }
+
+    /// Continue execution until the address is hit.
+    ///
+    /// Note that a breakpoint must not already exist for the target address, otherwise
+    /// this function will fail.
+    ///
+    /// This function will also fail if a pre-existing breakpoint is hit before the target
+    /// address is hit.
+    pub fn run_until(&self, address: impl Into<SymbolicAddress>, timeout: Duration) -> Result<()> {
+        let resolved = self.resolve(address)?;
+        log::info!("Run until {}", resolved);
+        self.jtag.set_breakpoint(resolved.resolution, true)?;
+        self.jtag.resume()?;
+        self.jtag.wait_halt(timeout)?;
+        self.expect_pc(resolved.resolution)?;
+        self.jtag.remove_breakpoint(resolved.resolution)?;
+        Ok(())
+    }
+
+    /// Excecute until the current function returns.
+    ///
+    /// This implementation does not use the debugging information from ELF files, and only use the RA register,
+    /// so it only works when RA has not been overriden, e.g. at the preamble of the function.
+    pub fn finish(&self, timeout: Duration) -> Result<()> {
+        let ra = self.read_reg(RiscvGpr::RA)?;
+        self.run_until(ra, timeout)
+    }
+
+    /// Call a function with the given arguments.
+    pub fn call(
+        &self,
+        address: impl Into<SymbolicAddress>,
+        args: &[u32],
+        timeout: Duration,
+    ) -> Result<(u32, u32)> {
+        // We're oblivious to the current context, so save all caller-saved registers.
+        // In addition, also save SP, just in case that we have to modify it to align.
+        const REGS_TO_SAVE: &[RiscvGpr] = &[
+            RiscvGpr::RA,
+            RiscvGpr::SP,
+            RiscvGpr::T0,
+            RiscvGpr::T1,
+            RiscvGpr::T2,
+            RiscvGpr::A0,
+            RiscvGpr::A1,
+            RiscvGpr::A2,
+            RiscvGpr::A3,
+            RiscvGpr::A4,
+            RiscvGpr::A5,
+            RiscvGpr::A6,
+            RiscvGpr::A7,
+            RiscvGpr::T3,
+            RiscvGpr::T4,
+            RiscvGpr::T5,
+            RiscvGpr::T6,
+        ];
+        let mut saved = [0; REGS_TO_SAVE.len()];
+        for (idx, gpr) in REGS_TO_SAVE.iter().copied().enumerate() {
+            saved[idx] = self.read_reg(gpr)?;
+        }
+
+        // Align the stack if it's not currently 16-byte aligned.
+        if saved[1] % 16 != 0 {
+            self.write_reg(RiscvGpr::SP, saved[1] & !15)?;
```

Comment:
Maybe add a comment to note that `saved[1] & !15` will round **down** so the function called will not overwrite the existing content of the stack?